jax/interpreters/pxla.py

```diff
@@ -345,30 +332,56 @@ def block_until_ready(self):
   @property
   def _value(self):
     if self._npy_value is None:
-      ids = self._ids()
+      # TODO(skye): remove this to avoid transferring replicated buffers?
       self.copy_to_host_async()
-      self._npy_value = self._collect([self.device_buffers[i].to_py() for i in ids])
+      self._npy_value = onp.empty(self.aval.shape, self.aval.dtype)
+      # TODO(skye): benchmark and possibly switch to a set (maybe with with
+      # cached hashable indices?)
+      already_copied_indices = []
+      for buf, idx in zip(self.device_buffers, self.logical_indices):
+        if idx in already_copied_indices:
+          continue
+        self._npy_value[idx] = buf.to_py()
+        already_copied_indices.append(idx)
     return self._npy_value
 
   def __getitem__(self, idx):
-    if self._npy_value is None and type(idx) is int:
-      ids = self._ids()
-      device_buffer = self.device_buffers[ids[idx]]
-      aval = ShapedArray(self.aval.shape[1:], self.aval.dtype)
-      return xla.DeviceArray(aval, None, lazy.array(aval.shape), device_buffer)
+    if self._npy_value is None and idx in self.logical_indices:
+      buf = self.device_buffers[self.logical_indices.index(idx)]
+      aval = ShapedArray(buf.shape().dimensions(), self.aval.dtype)
+      return xla.DeviceArray(aval, None, lazy.array(aval.shape), buf)
     else:
       return super(ShardedDeviceArray, self).__getitem__(idx)
 
-# This handler code is effectively dead because we in-lined it in shard_args for
-# performance reasons.
-def _shard_sharded_device_array(x, devices, assignments):
-  n = len(devices)
-  if n == len(x.device_buffers):
-    return (b if b.device() == devices[r] else b.copy_to_device(devices[r])
-            for r, b in enumerate(x.device_buffers))
-  else:
-    return (xla.device_put(x[assignments[r]], devices[r]) for r in range(n))
-shard_arg_handlers[ShardedDeviceArray] = _shard_sharded_device_array
+def _hashable_index(idx):
+  return tree_map(lambda x: (x.start, x.stop) if type(x) == slice else x,
```

Comment:
I think this code assumes that the slice has trivial stride; is that checked or documented anywhere?

Reply:
Good catch. I added this constraint to the SDA docstring and added a TODO to check it. (We could in theory allow strides, but I don't know why anyone would want that...)